lib/core/proxy.c

```diff
@@ -51,6 +52,19 @@ struct rp_generator_t {
     int *generator_disposed;
 };
 
+struct pipe_reader {
+    int fds[2];
+    h2o_linklist_t link;
+};
+
+struct pipe_list {
+    size_t len;
+    h2o_linklist_t anchor;
+};
+
+static __thread int pipe_list_init = 0;
+static __thread struct pipe_list pipe_list;
```

Comment:
> Maybe h2o_context_t::proxy is a such place.

IMHO: the limit being global but the implementation being private seems OK to me.

If you prefer that the `struct pipe_list` is moved to `h2o_context_t::proxy`, I can do that, but the side effect of this I think is that other code which has access to `h2o_context_t::proxy` will be exposed to the internals of pipe reuse which may not be necessary.

Can you let me know if you still prefer that `pipe_list` be relocated to `h2o_context_t::proxy` ?

Reply:
> IMHO: the limit being global but the implementation being private seems OK to me.

The problem is this limit being a property of an ordinary variable (that is not necessarily limited to one instance per process) controlling a process-wide state. Libh2o does not prohibit its users running two HTTP servers using different `h2o_globalconf_t`s.

> the side effect of this I think is that other code which has access to `h2o_context_t::proxy` will be exposed to the internals of pipe reuse which may not be necessary.

This is fine. There is indeed a trade-off here, but we do not try hard to hide implementation details. In case of the proxy module, we already expose pooled connections.

> Can you let me know if you still prefer that `pipe_list` be relocated to `h2o_context_t::proxy` ?

Thank you for asking. Yes, that is my preference. WDYT?

Reply:
OK, I've moved pipe_list in commit 610b83c9db0a3e67074899c93421e4a9af02c9c6.

I'll leave this conversation unresolved until you review it to ensure I did what you were thinking correctly. I've also moved initialization of the list to `h2o_context_init`.

Reply:
I've merged your PR with changes to this code that I think make this conversation resolved, so I'll mark this as resolved.